crawler/src/main.rs

```diff
@@ -0,0 +1,299 @@
+                Data::Object(genesis.header().clone()),
+            );
+
+            loop {
+                if node.node().num_connected() != 0 {
+                    info!(parent: node.node().span(), "sending out Pings");
+                    node.send_broadcast(ping_msg.clone());
+                }
+                tokio::time::sleep(Duration::from_secs(PING_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Spawns a task dedicated to peer maintenance.
+    pub fn update_peers(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            loop {
+                // Disconnect from peers we have just crawled.
+                for addr in node.known_network.addrs_to_disconnect() {
+                    node.node().disconnect(addr).await;
+                }
+
+                // Connect to peers we haven't crawled in a while.
+                for addr in node.known_network.addrs_to_connect() {
+                    if !node.node().is_connected(addr) {
+                        let _ = node.node().connect(addr).await;
+                    }
+                }
+
+                info!(parent: node.node().span(), "Crawling the netowrk for more peers; asking peers for their peers");
+                node.send_broadcast(ClientMessage::PeerRequest);
+                tokio::time::sleep(Duration::from_secs(PEER_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Starts the usual periodic activities of a crawler node.
+    pub fn run_periodic_tasks(&self) {
+        self.send_pings();
+        self.update_peers();
+    }
+}
+
+/// Inbound message processing logic for the test nodes.
+#[async_trait::async_trait]
+impl Reading for Crawler {
+    type Message = ClientMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        // FIXME: use the maximum message size allowed by the protocol or (better) use streaming deserialization.
+        let mut buf = [0u8; 8 * 1024];
+
+        reader.read_exact(&mut buf[..MESSAGE_LENGTH_PREFIX_SIZE])?;
+        let len = u32::from_le_bytes(buf[..MESSAGE_LENGTH_PREFIX_SIZE].try_into().unwrap()) as usize;
+
+        if len > buf.len() {
+            return Ok(None);
+        }
+
+        if reader.read_exact(&mut buf[..len]).is_err() {
+            return Ok(None);
+        }
+
+        match ClientMessage::deserialize(&buf[..len]) {
+            Ok(msg) => {
+                info!(parent: self.node().span(), "received a {} from {}", msg.name(), source);
+                Ok(Some(msg))
+            }
+            Err(e) => {
+                error!("a message from {} failed to deserialize: {}", source, e);
+                Err(io::ErrorKind::InvalidData.into())
+            }
+        }
+    }
+
+    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
+        // We may need to handle other messages as well.
+        match message {
```

Comment:
this logic is missing the recent pool-related variants

Reply:
Same in the test node, will add 👍